### PR TITLE
fix issue 17290 - DMD crash due to circular reference in function return type

### PR DESCRIPTION
**Pull request by ghost**
The fix is actually a generalization of what's been done in #9365.

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -3679,23 +3679,28 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
-
-        // by error the declaration can use itself in its type
-        if (vd)
-            vd.inuse++;
-
         if (!e.type)
             e.type = e.var.type;
         if (e.type && !e.type.deco)
+        {
+            Declaration decl = e.var.isDeclaration;
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d]: `auto`, parentheses

**Review by thewilsonator — APPROVED** [review]
Otherwise looks good.

**Comment by thewilsonator** [comment]
```
Test runnable/gdb15729.sh failed. The logged output:
warning: Error disabling address space randomization: Operation not permitted
test_results/runnable/gdb15729.gdb:2: Error in sourced command file:
During startup program exited normally.
==============================
```
Huh?

**Comment by Basile-z** [comment]
Unfortunately I don't have the privileges to relaunch the CircleCI tests myself, can you @thewilsonator  ?

**Comment by thewilsonator** [comment]
Magic! it appears to have fixed itself.

**Comment by Basile-z** [comment]
No magic here, the problem in #9404

**Pull request merged**